README.md
# tRPC Fetch-SSE Adapter

Because tRPC transmits data as JSON, sending (and subscribing to) individual Server-Sent event streams is not possible by default (which is handy if, say, you want to use SSE to send chunks of a ChatGPT response as they are generated). This adapter enables that functionality.

</br>

## Table of Contents

- [Usage](#usage)
  - [Adding the Adapter](#adding-the-adapter)
  - [Creating SSE Stream Procedures](#creating-sse-stream-procedures)
- [License](#license)
- [Author](#author)

</br>

## Usage

> See [the `trpc-sse-link` package](https://github.com/alecvision/trpc-sse-link) for the client-side link needed to consume SSE streams.

First, install the adapter:

```bash
npm install @alecvision/trpc-sse-adapter
```

There are two steps to implementing this adapter:

1. Add the adapter to your server and tell it which procedures are SSE streams
2. Create `subscription` procedures for your SSE streams

</br>

### Adding the Adapter

---

This adapter ONLY handles requests for SSE streams. Batching of SSE Stream requests is not supported. Creating an SSE stream is as simple as creating a subscription procedure, just as you would with WebSockets - but tRPC doesn't know the difference between a WebSocket and an SSE stream. You must tell it which procedures are SSE streams and handle them accordingly. For example, using Next.js:

```ts
import type { NextRequest } from "next/server";
import { fetchRequestHandler } from "@trpc/server/adapters/fetch";
import { sseRequestHandler } from "@alecvision/trpc-sse-adapter";
import { appRouter, createTRPCContext } from "../../../server";

const SSE_PROCEDURE_PATTERNS = [
  /ticker\.start$/,
  /chatgpt\.generate$/,
  /*
    prefixes/suffixes are an easy way to arbitrarily define SSE streams by giving
    them a special name (e.g. `myProcedure.stream_getSomeStreamingData`)
    */
  /^.*\.stream_\w+$/
];

// This MUST return the same value as is returned by the equivalent client-side function
function isStreamable(path: string) {
  return SSE_PROCEDURE_PATTERNS.some(regex => regex.test(path));
}

// Vercel only supports SSE on the edge runtime (WebSockets are not supported at all)
export const config = {
  runtime: "edge"
};

export default async function handler(req: NextRequest) {
  if (isStreamable(req.nextUrl.pathname)) {
    // Accepts a subset of the options for the fetch adapter
    return sseRequestHandler({
      endpoint: "/api/trpc",
      router: appRouter,
      req,
      createContext: createTRPCContext
    });
  }

  return fetchRequestHandler({
    endpoint: "/api/trpc",
    router: appRouter,
    req,
    createContext: createTRPCContext
  });
}

export default handler;
```

</br>

### Creating SSE Stream Procedures

---

```ts
import { observable } from "@trpc/server/observable";
import { OpenAI } from "openai-streams";
import { z } from "zod";
import { createTRPCRouter, publicProcedure } from ".";

export const chatRouter = createTRPCRouter({
  generate: publicProcedure
    .input(
      z.object({
        model: z.string(),
        messages: z.array(
          z.object({
            role: z.enum(["user", "system", "assistant"]),
            content: z.string()
          })
        ),
        temperature: z.number().nullish(),
        top_p: z.number().nullish(),
        frequency_penalty: z.number().nullish(),
        presence_penalty: z.number().nullish(),
        max_tokens: z.number().default(4096),
        n: z.number().nullish(),
        logit_bias: z
          .record(z.string(), z.number().min(-100).max(100))
          .nullish(),
        stop: z.array(z.string()).nullish(),
        user: z.string().nullish()
      })
    )
    .subscription(({ input }) => {
      return observable<string>(observer => {
        const abortController = new AbortController();

        void OpenAI("chat", input, {
          controller: abortController,
          apiKey: process.env.OPEN_AI_API_KEY,
          onParse(token) {
            observer.next(token);
          },
          onDone() {
            observer.complete();
          }
        }).catch(err => {
          observer.error(err);
        });

        return () => {
          abortController.abort();
          observer.complete();
        };
      });
    })
});
```

</br>

## License

ISC License (ISC)

</br>

## Author

Alec Helmturner


CHANGELOG.md
# Changelog

## [0.0.2](https://github.com/AlecVision/trpc-sse-adapter/compare/trpc-sse-adapter-v0.0.1...trpc-sse-adapter-v0.0.2) (2023-06-06)


### Features

* update API to use `withSSE` wrapper ([8f2ff27](https://github.com/AlecVision/trpc-sse-adapter/commit/8f2ff2754fd47a4e14f5b186390200053136a14d))

## 0.0.1 (2023-06-03)


### Features

* initial commit ([566088f](https://github.com/AlecVision/trpc-sse-adapter/commit/566088f614bf7055d560a2b1284ee15718f5225c))


src/toss.ts
import { TRPCError } from "@trpc/server";
import {
  TRPC_ERROR_CODE_KEY,
  TRPC_ERROR_CODE_NUMBER,
  TRPC_ERROR_CODES_BY_NUMBER
} from "@trpc/server/rpc";

/**
 * Like throwing, but shorter.
 * @param message - error message
 * @param code - TRPC error code (you can lookup numeric codes with `TRPC_ERROR_CODES_BY_NUMBER[`(error code)`]`)
 * @param cause - optional cause (usually this will be the argument passed to `catch`)
 * @throws {TRPCError}
 * @see TRPC_ERROR_CODES_BY_NUMBER:
 * - "-32700": "PARSE_ERROR";
 * - "-32600": "BAD_REQUEST";
 * - "-32603": "INTERNAL_SERVER_ERROR";
 * - "-32001": "UNAUTHORIZED";
 * - "-32003": "FORBIDDEN";
 * - "-32004": "NOT_FOUND";
 * - "-32005": "METHOD_NOT_SUPPORTED";
 * - "-32008": "TIMEOUT";
 * - "-32009": "CONFLICT";
 * - "-32012": "PRECONDITION_FAILED";
 * - "-32013": "PAYLOAD_TOO_LARGE";
 * - "-32022": "UNPROCESSABLE_CONTENT";
 * - "-32029": "TOO_MANY_REQUESTS";
 * - "-32099": "CLIENT_CLOSED_REQUEST
 */
export function toss(
  message: string,
  code: TRPC_ERROR_CODE_KEY | TRPC_ERROR_CODE_NUMBER,
  cause?: unknown
): never {
  if (typeof code === "number") code = TRPC_ERROR_CODES_BY_NUMBER[code];
  const err = new TRPCError({
    code,
    message,
    cause
  });
  console.error(err);
  throw err;
}

//@ts-expect-error - Vite handles this import.meta check
if (import.meta.vitest) {
  //@ts-expect-error - Vite handles this top-level await
  const [{ describe }] = await Promise.all([import("vitest")]);
  describe("toss", it => {
    it("should work", async ({ expect }) => {
      expect(true).toBe(true);
    });
  });
}


src/index.ts
import { AnyRouter } from "@trpc/server";
import { FetchHandlerRequestOptions } from "@trpc/server/adapters/fetch";
import { HTTPRequest } from "@trpc/server/http";
import { ResponseMetaFn, resolveSSEResponse } from "./resolveSSEResponse";

export type SSEHandlerRequestOptions<TRouter extends AnyRouter> = Omit<
  FetchHandlerRequestOptions<TRouter>,
  "responseMeta"
> & {
  responseMeta?: ResponseMetaFn<TRouter>;
};

export async function sseRequestHandler<TRouter extends AnyRouter>(
  opts: SSEHandlerRequestOptions<TRouter>
): Promise<Response> {
  const {
    endpoint,
    router,
    //TODO: batching,
    onError,
    responseMeta,
    req,
    createContext
  } = opts;

  const resHeaders = new Headers();

  const { pathname, searchParams } = new URL(req.url);
  const path = pathname.slice(endpoint.length + 1);
  const isBodyJSON = req.headers.get("Content-Type") === "application/json";
  const httpRequest: HTTPRequest = {
    query: searchParams,
    method: req.method,
    headers: Object.fromEntries(req.headers),
    body: isBodyJSON ? await req.text() : ""
  };

  const result = await resolveSSEResponse({
    req: httpRequest,
    path,
    router,
    // only set if we have a value
    //TODO:    ...(batching && { batching }),
    ...(responseMeta && { responseMeta }),
    createContext: async () => createContext?.({ req, resHeaders }),
    onError: o => void onError?.({ ...o, req })
  });

  for (const [key, value] of Object.entries(result.headers ?? {})) {
    /* istanbul ignore if -- @preserve */
    if (typeof value === "string") resHeaders.set(key, value);
    if (Array.isArray(value)) value.forEach(v => resHeaders.append(key, v));
  }

  return new Response(result.body, {
    status: result.status,
    headers: resHeaders
  });
}

/* c8 ignore start */
//@ts-expect-error - Vite handles this import.meta check
if (import.meta.vitest) {
  //@ts-expect-error - Vite handles this top-level await
  const [{ describe }] = await Promise.all([import("vitest")]);
  describe("sseRequestHandler", it => {
    it("should work", async ({ expect }) => {
      expect(true).toBe(true);
    });
  });
}


src/sseHandler.ts
import { resolveSSEResponse } from "./resolveSSEResponse";
import type { AnyRouter } from "@trpc/server";
import type { FetchHandlerRequestOptions } from "@trpc/server/adapters/fetch";

export async function sseRequestHandler<TRouter extends AnyRouter>(
  opts: FetchHandlerRequestOptions<TRouter>
): Promise<Response> {
  const { endpoint, req, router, responseMeta, onError, createContext } = opts;

  const resHeaders = new Headers();

  const { pathname, searchParams } = new URL(req.url);
  const path = pathname.slice(endpoint.length + 1);
  const isBodyJSON = req.headers.get("Content-Type") === "application/json";
  const httpRequest = {
    query: searchParams,
    method: req.method,
    headers: Object.fromEntries(req.headers),
    body: isBodyJSON ? await req.text() : ""
  };

  //TODO: Clean this up after fetchRequestHandler is updated
  const result = await resolveSSEResponse({
    req: httpRequest,
    path,
    router,
    ...(responseMeta && { responseMeta }),
    createContext: async () => createContext?.({ req, resHeaders }),
    onError: o => void onError?.({ ...o, req })
  });

  for (const [key, value] of Object.entries(result.headers ?? {})) {
    if (typeof value === "string") resHeaders.set(key, value);
    if (Array.isArray(value)) value.forEach(v => resHeaders.append(key, v));
  }

  return new Response(result.body, {
    status: result.status,
    headers: resHeaders
  });
}

/* c8 ignore start */

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

//@ts-expect-error - Vite handles this import.meta check
if (import.meta.vitest) {
  //@ts-expect-error - Vite handles this top-level await
  const [{ describe }] = await Promise.all([import("vitest")]);
  describe("sseRequestHandler", it => {
    it("should work", async ({ expect }) => {
      expect(true).toBe(true);
    });
  });
}

/* c8 ignore end */


src/resolveSSEResponse.ts
import type { HTTPHeaders } from "@trpc/client";
import {
  type AnyProcedure,
  type ProcedureRouterRecord,
  TRPCError,
  getTRPCErrorFromUnknown,
  inferRouterContext,
  inferRouterError,
  type AnyRouter,
  type CombinedDataTransformer,
  type Maybe,
  type MaybePromise,
  type ProcedureType
} from "@trpc/server";
import {
  getHTTPStatusCode,
  type ResponseMeta,
  type HTTPRequest,
  type HTTPBaseHandlerOptions
} from "@trpc/server/http";
import { isObservable } from "@trpc/server/observable";
import {
  type TRPCResponse,
  type TRPCResponseMessage,
  TRPC_ERROR_CODES_BY_KEY
} from "@trpc/server/rpc";
import { getErrorShape, transformTRPCResponse } from "@trpc/server/shared";
import { toss } from "./toss";

const HTTP_METHOD_PROCEDURE_TYPE_MAP: Record<
  string,
  ProcedureType | undefined
> = {
  GET: "query",
  PATCH: "subscription",
  POST: "mutation"
};

const fallbackContentTypeHandler = {
  getInputs: getJsonContentTypeInputs
};

export async function resolveSSEResponse<TRouter extends AnyRouter>(
  opts: ResolveSSERequestOptions<TRouter, HTTPRequest>
): Promise<HTTPResponse> {
  const {
    router,
    req,
    createContext,
    onError,
    path,
    responseMeta,
    error,
    preprocessedBody = false,
    //TODO: batching: _batching,
    contentTypeHandler = fallbackContentTypeHandler
  } = opts;

  // can be used for lambda warmup
  if (req.method === "HEAD") return { status: 204 };

  const method = req.method;
  const type = HTTP_METHOD_PROCEDURE_TYPE_MAP[method] ?? "unknown";
  let ctx: inferRouterContext<TRouter> | undefined = undefined;

  //TODO: Const isBatchCall = !!req.query.get("batch");
  const paths: string[] = /* isBatchCall ? path.split(",") : */ [path];

  const getInputs = contentTypeHandler.getInputs.bind(contentTypeHandler, {
    //TODO: isBatchCall,
    req,
    router,
    preprocessedBody
  });

  type TRouterError = inferRouterError<TRouter>;
  type TRouterResponse = TRPCResponse<unknown, TRouterError>;

  // TODO: support batching? Is it possible to combine this with https://trpc.io/docs/links/httpBatchStreamLink ?

  try {
    if (error) throw error;
    //TODO: if (isBatchCall) toss(`Batching SSE requests is not supported`, -32022);
    if (type !== "subscription") toss(`Unexpected method: ${method}`, -32005);

    ctx = await createContext();

    const inputs = await getInputs();

    const rawResults = await Promise.all(
      paths.map(async (path, index) => {
        const input = inputs[index];

        try {
          const output = await callProcedure({
            procedures: router._def.procedures,
            path,
            rawInput: input,
            ctx,
            type
          });
          return {
            input,
            path,
            data: output
          };
        } catch (cause) {
          const error = getTRPCErrorFromUnknown(cause);

          onError?.({ error, path, input, ctx, type, req });
          return {
            input,
            path,
            error
          };
        }
      })
    );

    const errors = rawResults.flatMap(obj => (obj.error ? [obj.error] : []));
    const resultEnvelopes = rawResults.map(
      ({ input, path, error, data }): TRouterResponse => {
        if (!error) return { result: { data } };
        return {
          error: getErrorShape({
            config: router._def._config,
            error,
            type,
            path,
            input,
            ctx
          })
        };
      }
    );

    const result =
      /*TODO: isBatchCall ? resultEnvelopes :  */ resultEnvelopes[0]!;

    return endResponse(result, errors);
  } catch (cause) {
    // we get here if
    // - batching is called when it's not enabled
    // - `createContext()` throws
    // - post body is too large
    // - input deserialization fails
    // - `errorFormatter` return value is malformed
    const error = getTRPCErrorFromUnknown(cause);

    onError?.({
      error,
      path: undefined,
      input: undefined,
      ctx,
      type,
      req
    });
    return endResponse(
      {
        error: getErrorShape({
          config: router._def._config,
          error,
          type,
          path: undefined,
          input: undefined,
          ctx
        })
      },
      [error]
    );
  }

  function endResponse(
    untransformedJSON: TRouterResponse | TRouterResponse[],
    errors: TRPCError[]
  ): HTTPResponse {
    let status = getHTTPStatusCode(untransformedJSON);

    const headers: HTTPHeaders = {};

    const meta =
      responseMeta?.({
        ...(paths && { paths }), // only set if we have paths
        ctx,
        type,
        data: Array.isArray(untransformedJSON)
          ? untransformedJSON
          : [untransformedJSON],
        errors
      }) ?? {};

    status = meta?.status ?? status;

    for (const [key, value] of Object.entries(meta.headers ?? {})) {
      headers[key] = value;
    }

    if (type === "subscription") {
      untransformedJSON = [untransformedJSON].flat()[0]!; // assert there is only one result and unwrap it

      if ("error" in untransformedJSON) {
        console.error(
          "ERROR IN SUBSCRIPTION: ",
          JSON.stringify(untransformedJSON, null, 2)
        );
        return {
          body: JSON.stringify(untransformedJSON),
          status,
          headers: {
            ...headers,
            "Content-Type": "text/event-stream",
            Connection: "keep-alive",
            // see https://github.com/vercel/next.js/issues/9965#issuecomment-587355489
            "Cache-Control": "no-cache, no-transform",
            // avoid buffering by reverse proxies (e.g. nginx)
            "X-Accel-Buffering": "no"
          }
        };
      }

      const observable = untransformedJSON.result.data;

      const rpcMeta = {
        id: `${untransformedJSON.id}`,
        jsonrpc: "2.0" as const
      };

      if (!isObservable(observable)) {
        toss(`Subscription ${path} did not return an observable`, -32603);
      }

      let count = 0;
      const encoder = new TextEncoder();

      /**
       *
       * @param controller
       * @param input "START" | "STOP" sends a "started" | "stopped" event. Any other input is sent as data
       */
      function sendEvent(
        controller: ReadableStreamDefaultController<Uint8Array>,
        input: unknown
      ): void {
        let result;
        switch (input) {
          case "START":
            result = { type: "started" as const };
            break;
          case "STOP":
            result = { type: "stopped" as const };
            break;
          default:
            result = { data: input, type: "data" as const };
            break;
        }

        const envelope: TRPCResponseMessage = { result, ...rpcMeta };
        const json = transformTRPCResponse(router._def._config, envelope);

        const eventId = `id:${rpcMeta.id}/${count++}`;
        const eventData = `data:${JSON.stringify(json)}`;
        const event = `${eventData}\n${eventId}\n\n`;

        const bytes = encoder.encode(event);
        controller.enqueue(bytes);
      }

      const body = new ReadableStream<Uint8Array>({
        async start(controller) {
          const subscription = observable.subscribe({
            next: data => {
              if (count === 0) sendEvent(controller, "START");
              sendEvent(controller, data);
            },
            complete: () => {
              sendEvent(controller, "STOP");
              controller.close();
            },
            error: async err => {
              const {
                code: status_code,
                message,
                name
              } = getTRPCErrorFromUnknown(err);
              const input = (await getInputs())[paths.indexOf(path)];
              const data = { path, input, type, error: name };
              const code = TRPC_ERROR_CODES_BY_KEY[status_code];
              const envelope = { error: { code, message, data }, ...rpcMeta };
              const json = transformTRPCResponse(router._def._config, envelope);
              //TODO: test this by throwing error from procedure
              controller.error(json);
            }
          });

          return () => {
            subscription.unsubscribe();
          };
        }
      });

      return {
        body,
        status,
        headers
      };
    }

    const transformedJSON = transformTRPCResponse(
      router._def._config,
      untransformedJSON
    );

    const body = JSON.stringify(transformedJSON);

    return {
      body,
      status,
      headers: {
        ...headers,
        "Content-Type": "application/json"
      }
    };
  }
}

function getRawProcedureInputOrThrow(opts: {
  req: HTTPRequest;
  preprocessedBody: boolean;
}) {
  const { req, preprocessedBody } = opts;
  try {
    if (req.method === "GET") {
      if (!req.query.has("input")) {
        return undefined;
      }
      const raw = req.query.get("input");
      return JSON.parse(raw!);
    }
    if (!preprocessedBody && typeof req.body === "string") {
      // A mutation with no inputs will have req.body === ''
      return req.body.length === 0 ? undefined : JSON.parse(req.body);
    }
    return req.body;
  } catch (err) {
    throw new TRPCError({
      code: "PARSE_ERROR",
      cause: err instanceof Error ? err : undefined
    });
  }
}

function deserializeInputValue(
  rawValue: unknown,
  transformer: CombinedDataTransformer
) {
  return typeof rawValue === "undefined"
    ? undefined
    : transformer.input.deserialize(rawValue);
}

function getJsonContentTypeInputs(
  opts: Parameters<GetInputs>[0]
): ReturnType<GetInputs> {
  const rawInput = getRawProcedureInputOrThrow(opts);
  const transformer = opts.router._def._config.transformer;

  //TODO:if (!opts.isBatchCall) {
  return {
    0: deserializeInputValue(rawInput, transformer)
  };
  /*TODO:   }

  assertValidBatchInput(rawInput);

  const input: typeof rawInput = {};
  for (const key in rawInput) {
    const k = key as `${number}`;
    input[k] = deserializeInputValue(rawInput[k], transformer);
  }

  return input;
  */
}

function callProcedure(opts: {
  ctx: unknown;
  rawInput: unknown;
  input?: unknown;
  path: string;
  type: ProcedureType;
  procedures: ProcedureRouterRecord;
}) {
  const { type, path, procedures } = opts;

  if (!(path in procedures) || !procedures[path]?._def[type]) {
    toss(`No "${type}"-procedure on path "${path}"`, -32004);
  }

  const procedure = procedures[path] as AnyProcedure;
  return procedure(opts);
}

interface ResolveSSERequestOptions<
  TRouter extends AnyRouter,
  TRequest extends HTTPRequest
> extends HTTPBaseHandlerOptions<TRouter, TRequest> {
  createContext: () => Promise<inferRouterContext<TRouter>>;
  req: TRequest;
  path: string;
  error?: Maybe<TRPCError>;
  contentTypeHandler?: BaseContentTypeHandler<any>;
  preprocessedBody?: boolean;
  responseMeta?: ResponseMetaFn<TRouter>;
}

interface HTTPResponse {
  status: number;
  headers?: HTTPHeaders;
  body?: string | ReadableStream<Uint8Array>;
}

type GetInputs = (opts: {
  req: HTTPRequest;
  //TODO:  isBatchCall: boolean;
  router: AnyRouter;
  preprocessedBody: boolean;
}) => MaybePromise<Record<number, unknown>>;

type BodyResult =
  | {
      ok: true;
      data: unknown;
      /**
       * If the HTTP handler has already parsed the body
       */
      preprocessed: boolean;
    }
  | { ok: false; error: TRPCError };

type BaseContentTypeHandler<TOptions> = {
  isMatch(opts: TOptions): boolean;
  getBody: (opts: TOptions) => MaybePromise<BodyResult>;
  getInputs: GetInputs;
};

export type ResponseMetaFn<TRouter extends AnyRouter> = (opts: {
  data: TRPCResponse<unknown, inferRouterError<TRouter>>[];
  ctx?: inferRouterContext<TRouter>;
  /**
   * The different tRPC paths requested
   **/
  paths?: string[];
  type: ProcedureType | "unknown";
  errors: TRPCError[];
}) => ResponseMeta;

/*
//TODO
function assertValidBatchInput(
  x: unknown,
): asserts x is Record<`${number}`, unknown> {
  if (typeof x === "object" && x && !Array.isArray(x)) return;
  toss('"input" must be an object when doing a batch call', -32600);
} */

/* c8 ignore start */
//@ts-expect-error - Vite handles this import.meta check
if (import.meta.vitest) {
  //@ts-expect-error - Vite handles this top-level await
  const [{ describe }] = await Promise.all([import("vitest")]);
  describe("sseRequestResolver", it => {
    it("should work", async ({ expect }) => {
      expect(true).toBe(true);
    });
  });
}
